dotnet/src/Azure.Iot.Operations.Protocol/CloudEvent-Architecture.md

```diff
@@ -0,0 +1,225 @@
+        +ProtocolCloudEvent? GetProtocolCloudEvent() internal
+    }
+    
+    class OutgoingTelemetryMetadata {
+        +ProtocolCloudEvent? CloudEvent
+        +Dictionary~string,string~ UserData
+    }
+    
+    class IncomingTelemetryMetadata {
+        +ProtocolCloudEvent GetCloudEvent()
+        +Dictionary~string,string~ UserData
+    }
+    
+    class CommandRequestMetadata {
+        +ProtocolCloudEvent? CloudEvent
+        +Dictionary~string,string~ UserData
+    }
+    
+    class CommandResponseMetadata {
+        +ProtocolCloudEvent? CloudEvent
+        +Dictionary~string,string~ UserData
+    }
+    
+    class TelemetrySender {
+        +SendTelemetryAsync(payload, metadata)
+    }
+    
+    class CommandInvoker {
+        +InvokeAsync(payload, metadata)
+    }
+    
+    class CommandExecutor {
+        +StartAsync()
+    }
+
+    MqttApplicationMessage --> CloudEvent : "Generic MQTT<br>(user-controlled)"
+    MqttApplicationMessage --> ProtocolCloudEvent : "Protocol use<br>(SDK-controlled)"
+    
+    OutgoingTelemetryMetadata --> ProtocolCloudEvent
+    IncomingTelemetryMetadata --> ProtocolCloudEvent
+    CommandRequestMetadata --> ProtocolCloudEvent
+    CommandResponseMetadata --> ProtocolCloudEvent
+    
+    TelemetrySender --> OutgoingTelemetryMetadata
+    CommandInvoker --> CommandRequestMetadata
+    CommandInvoker --> CommandResponseMetadata
+    CommandExecutor --> CommandRequestMetadata
+    CommandExecutor --> CommandResponseMetadata
+    
+    note for CloudEvent "User controls Type and DataContentType<br>Requires explicit type in constructor<br>For generic MQTT scenarios"
+    
+    note for ProtocolCloudEvent "SDK controls Type and DataContentType<br>Type defaults to 'ms.aio.telemetry'<br>For protocol-specific scenarios"
+```
+
+## Usage Scenarios
+
+### Generic MQTT Usage (CloudEvent)
+
+```mermaid
+sequenceDiagram
+    participant User as User Code
+    participant MQTT as MqttApplicationMessage
+    participant CE as CloudEvent
+    
+    User->>CE: new CloudEvent(source, "my.custom.type")
+    User->>CE: Set properties (Id, Time, etc.)
+    User->>MQTT: SetCloudEvent(cloudEvent)
+    MQTT->>MQTT: Serialize to User Properties
+    
+    Note over User,MQTT: User has full control over <br> Type and DataContentType
+```
+
+### Protocol Telemetry Usage (ProtocolCloudEvent)
+
+```mermaid
+sequenceDiagram
+    participant User as User Code
+    participant Meta as OutgoingTelemetryMetadata
+    participant PCE as ProtocolCloudEvent
+    participant Sender as TelemetrySender
+    participant MQTT as MqttApplicationMessage
+    
+    User->>PCE: new ProtocolCloudEvent(source)
+    Note over PCE: Type = "ms.aio.telemetry" (default)
+    User->>PCE: Set optional properties (Subject, DataSchema)
+    User->>Meta: metadata.CloudEvent = pce
+    User->>Sender: SendTelemetryAsync(payload, metadata)
+    Sender->>PCE: Set DataContentType (from serializer)
+    Note over Sender: SDK controls Type and DataContentType
```

Comment:
this is only true for the sending direction. When receiving, this is treated the same as the generic scenario